### PR TITLE
[perf] Change stable hasher to Blake3

### PR DESCRIPTION
**Pull request by Urgau**
Based on https://github.com/rust-lang/rustc-stable-hash/compare/main...Urgau:rustc-stable-hash:blake3

cc @michaelwoerister
r? ghost

### CODE REVIEW

File: compiler/rustc_data_structures/src/fingerprint.rs [hunk 1]
```diff
@@ -158,8 +158,12 @@ impl FromStableHash for Fingerprint {
     type Hash = StableHasherHash;
 
     #[inline]
-    fn from(StableHasherHash([_0, _1]): Self::Hash) -> Self {
-        Fingerprint(_0, _1)
+    fn from(hash: Self::Hash) -> Self {
+        let bytes = hash.as_bytes();
+        Fingerprint(
+            u64::from_ne_bytes(bytes[0..8].try_into().unwrap()),
+            u64::from_ne_bytes(bytes[8..16].try_into().unwrap()),
```
Comment (michaelwoerister) [inline compiler/rustc_data_structures/src/fingerprint.rs:165]: Converting from bytes needs to use a fixed endianess. Blake3 encodes the hash as little-endian it seems: https://github.com/BLAKE3-team/BLAKE3/blob/acaeabeff3e051d5396e91732b1fd2f0b2641402/src/lib.rs#L449

**Comment by Urgau** [comment]
I have no idea about what's happening. stage1 and stage2 seems fine, but directly after Bolt-ing LLVM we `SIGSEGV` instant on all the benchmarks after that.

<details>

> 2024-07-15T12:40:31.1170768Z Executing benchmark bitmaps-3.1.0 (1/8)
2024-07-15T12:40:31.1172991Z Preparing bitmaps-3.1.0
2024-07-15T12:40:31.1278780Z [2024-07-15T12:40:31Z INFO  collector::compile::execute] run_rustc with incremental=false, profile=Check, scenario=None, patch=None, backend=Llvm, phase=dependencies
2024-07-15T12:40:31.1281263Z [2024-07-15T12:40:31Z INFO  collector::compile::execute] run_rustc with incremental=false, profile=Debug, scenario=None, patch=None, backend=Llvm, phase=dependencies
2024-07-15T12:40:31.1283264Z [2024-07-15T12:40:31Z INFO  collector::compile::execute] run_rustc with incremental=false, profile=Opt, scenario=None, patch=None, backend=Llvm, phase=dependencies
2024-07-15T12:40:31.1392313Z [2024-07-15T12:40:31Z DEBUG collector::compile::execute] cd "/tmp/.tmplRWv8p" && CARGO="/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" CARGO_INCREMENTAL="0" CARGO_MAKEFLAGS="-j --jobserver-fds=9,10 --jobserver-auth=9,10" RUSTC="/tmp/tmp-multistage/opt-artifacts/rustc-perf/target/debug/rustc-fake" RUSTC_BOOTSTRAP="1" RUSTC_REAL="/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "path+file:///tmp/.tmplRWv8p#bitmaps@3.1.0" "--" "--skip-this-rustc"
2024-07-15T12:40:31.1400664Z [2024-07-15T12:40:31Z DEBUG collector::compile::execute] cd "/tmp/.tmp1vgxwS" && CARGO="/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" CARGO_INCREMENTAL="0" CARGO_MAKEFLAGS="-j --jobserver-fds=9,10 --jobserver-auth=9,10" RUSTC="/tmp/tmp-multistage/opt-artifacts/rustc-perf/target/debug/rustc-fake" RUSTC_BOOTSTRAP="1" RUSTC_REAL="/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "path+file:///tmp/.tmp1vgxwS#bitmaps@3.1.0" "--profile" "check" "--" "--skip-this-rustc"
2024-07-15T12:40:31.1406599Z [2024-07-15T12:40:31Z DEBUG collector::compile::execute] cd "/tmp/.tmpNZVCp2" && CARGO="/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" CARGO_INCREMENTAL="0" CARGO_MAKEFLAGS="-j --jobserver-fds=9,10 --jobserver-auth=9,10" RUSTC="/tmp/tmp-multistage/opt-artifacts/rustc-perf/target/debug/rustc-fake" RUSTC_BOOTSTRAP="1" RUSTC_REAL="/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "path+file:///tmp/.tmpNZVCp2#bitmaps@3.1.0" "--release" "--" "--skip-this-rustc"
2024-07-15T12:40:31.8870331Z Finished benchmark bitmaps-3.1.0 (1/8)
2024-07-15T12:40:31.8871538Z collector error: Failed to profile 'bitmaps-3.1.0' with Eprintln, recorded: expected success, got exit status: 101
2024-07-15T12:40:31.8872322Z 
2024-07-15T12:40:31.8872829Z stderr=error: failed to run `rustc` to learn about target-specific information
2024-07-15T12:40:31.8873407Z 
2024-07-15T12:40:31.8873556Z Caused by:
2024-07-15T12:40:31.8876116Z   process didn't exit successfully: `/tmp/tmp-multistage/opt-artifacts/rustc-perf/target/debug/rustc-fake - --crate-name ___ --print=file-names --crate-type bin --crate-type rlib --crate-type dylib --crate-type cdylib --crate-type staticlib --crate-type proc-macro --print=sysroot --print=split-debuginfo --print=crate-name --print=cfg` (signal: 11, SIGSEGV: invalid memory reference)
2024-07-15T12:40:31.8878432Z   --- stderr
2024-07-15T12:40:31.8878845Z   error: rustc interrupted by SIGSEGV, printing backtrace
2024-07-15T12:40:31.8879478Z 
2024-07-15T12:40:31.8880246Z   /checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/../lib/librustc_driver-cc756da8c2cf89a3.so(+0x96f9ec7)[0x7fdd1eb26ec7]
2024-07-15T12:40:31.8881208Z   /lib64/libpthread.so.0(+0xf630)[0x7fdd14e14630]
2024-07-15T12:40:31.8882252Z   /checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/../lib/librustc_driver-cc756da8c2cf89a3.so(+0x4e102a2)[0x7fdd1a23d2a2]
2024-07-15T12:40:31.8883019Z 
2024-07-15T12:40:31.8883421Z   note: we would appreciate a report at https://github.com/rust-lang/rust
2024-07-15T12:40:31.8884398Z   help: you can increase rustc's stack size by setting RUST_MIN_STACK=16777216
2024-07-15T12:40:31.8884888Z 
2024-07-15T12:40:31.8884894Z 
2024-07-15T12:40:31.8885000Z  stdout=
2024-07-15T12:40:31.8885256Z 

</details>

**Comment by michaelwoerister** [comment]
Is it possible that the Blake3 crate uses some low-level feature that BOLT does not support?

**Comment by Urgau** [comment]
We are seeing regressions up to 122% for secondary benchmarks and up to 62% in primary benchmarks with a mean 26.2% regressions.

Looking at the [BOLT-ing of LLVM](https://github.com/rust-lang/rust/pull/94381#issuecomment-1273109432) alone (2 years ago) it didn't even have a impact on instructions directly, granted many things have changed; and while BOLT-ing [`librustc_driver.so`](https://github.com/rust-lang/rust/pull/116352#issuecomment-1763328251) did have a impact on instructions around 2.3% mean. It's no where near the 26.2% regressions we are seeing here.

Those regressions seems to indicate that Blake 3 is not a suitable replacement for our SipHasher for stable hashing of small structs. It might perform much better with better inputs, at least that's what is being suggested with [their own benchmarks](https://github.com/BLAKE3-team/BLAKE3/blob/master/media/speed.svg) showing a 16Kib input.

In general it seems like Blake 3 is throughput-optimized, while for stable hashing, latency is the most important. It even has a 1024-bit fixed-input buffer, which probably explains why max-rss was so dearly affected.

**Comment by lqd** [comment]
To have more up to date results than 2 years ago, as bolt has also evolved since then, you could also do a perf run with only the commit disabling it.

**Comment by Nilstrieb** [comment]
https://jszym.com/blog/short_input_hash/

**Comment by Urgau** [comment]
> https://jszym.com/blog/short_input_hash/

Based on the results above and this article, I'm trying Blake2s in https://github.com/rust-lang/rust/pull/127809.

**Comment by Urgau** [comment]
Given the results above (even taking a considerable BOLT margin), Blake3 is probably a dead-end for our use-case.

Closing.

**Pull request closed without merging**